### ISSUE TITLE
Bug regarding SMTP and bare LF's/CRLF's

### ISSUE THREAD

**Issue by FabulousGee** [hmailserver/hmailserver#357]
I was going through a few options lately and activated the RFC compliance to block emails containing bare LF's. But as soon as I did this, some emails got blocked. One issue was resolved on my end (a small script of mine that wasn't configured properly) but one issue still persists: the users can not use our webmailer anymore! First I thought it might be an issue on their side but I came to the conclusion that the fault seems to be on hMailserver side - though I can not see why this might be happening. The behaviour is reproducible, though.

This is the communication I logged (sending email via RainLoop webmailer):
```
TAG1 CAPABILITY
TAG2 LOGIN "xxx" "xxx"
EHLO mail.xxx.de
AUTH LOGIN
xxx=
xxx=
MAIL FROM:<xxx@xxx.de>
RCPT TO:<xxx@xxx.de>
DATA
MIME-Version: 1.0
Date: Sun, 29 Nov 2020 22:02:26 +0000
Content-Type: multipart/alternative;
 boundary="--=_RainLoop_472_915006015.1606687346"
From: "xxx - xxx"
 <xxx@xxx.de>
Message-ID: <50ba42xxxx1fd13xxxxfc5768f79xxxx@xxx.de>
Subject: TESTemail
To: xxx@xxx.de
X-Originating-IP: xxx.xx.xx.xxx


----=_RainLoop_472_915006015.1606687346
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: quoted-printable

Das ist nur ein Test=0A=0Aein Test=0Amit verschiedenen Umbr=C3=BCchen=0A=
=0A	Testtext 12345=0A123421345=0A12345=0A13245=0A32145=0Aasd=0Ag=0Aasdfg=
=0Asdffg

----=_RainLoop_472_915006015.1606687346
Content-Type: text/html; charset="utf-8"
Content-Transfer-Encoding: quoted-printable

<!DOCTYPE html><html><head><meta http-equiv=3D"Content-Type" content=3D"t=
ext/html; charset=3Dutf-8" /></head><body><div data-html-editor-font-wrap=
per=3D"true" style=3D"font-family: arial, sans-serif; font-size: 13px;"><=
div><div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><d=
iv><div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><di=
v><div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><div=
><div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><p>Da=
s ist nur ein Test<br><br>ein Test<br>mit verschiedenen Umbr=C3=BCchen</p=
><p>Testtext 12345<br>123421345<br>12345<br>13245<br>32145<br>asd<br>g<br=
>asdfg<br>sdffg</p><signature></signature></div></div></div></div></div><=
/div></div></div></div></div></div></div></div></body></html>

----=_RainLoop_472_915006015.1606687346--
.
TAG3 LOGOUT
```

As you can see in the image, it's all CRLF, so no bare LF's seem to be present:
https://abload.de/img/crlfpyk42.png

This is the according hMailserver log:
```
"DEBUG"	8660	"2020-11-29 23:02:26.662"	"TCP connection started for session 1304"
"DEBUG"	8660	"2020-11-29 23:02:26.662"	"Performing SSL/TLS handshake for session 1304. Verify certificate: False"
"SMTPD"	8660	1304	"2020-11-29 23:02:26.724"	"xx.xxx.xxx.xxx"	"SENT: 220 server1.xxx.de"
"SMTPD"	8660	1304	"2020-11-29 23:02:26.724"	"xx.xxx.xxx.xxx"	"RECEIVED: EHLO mail.xxx.de"
"SMTPD"	8660	1304	"2020-11-29 23:02:26.724"	"xx.xxx.xxx.xxx"	"SENT: 250-xxx.de[nl]250-SIZE 209715200[nl]250-AUTH LOGIN[nl]250 HELP"
"SMTPD"	4988	1304	"2020-11-29 23:02:26.740"	"xx.xxx.xxx.xxx"	"RECEIVED: AUTH LOGIN"
"SMTPD"	4988	1304	"2020-11-29 23:02:26.740"	"xx.xxx.xxx.xxx"	"SENT: 334 xxx"
"SMTPD"	8988	1304	"2020-11-29 23:02:26.740"	"xx.xxx.xxx.xxx"	"RECEIVED: xxx="
"SMTPD"	8988	1304	"2020-11-29 23:02:26.740"	"xx.xxx.xxx.xxx"	"SENT: 334 xxx"
"SMTPD"	8660	1304	"2020-11-29 23:02:26.755"	"xx.xxx.xxx.xxx"	"RECEIVED: ***"
"DEBUG"	8660	"2020-11-29 23:02:26.755"	"Executing event OnClientValidatePassword"
"ERROR"	6324	"2020-11-29 23:02:27.052"	"Severity: 2 (High), Code: HM5032, Source: DALConnection::Execute, Description: MySQL: Incorrect string value: '\xF0\x9F\x92\xA5 B...' for column 'metadata_subject' at row 1 (Additional info: INSERT INTO hm_message_metadata (metadata_accountid, metadata_folderid, metadata_messageid, metadata_dateutc, metadata_from, metadata_subject, metadata_to, metadata_cc) VALUES (17, 55, 86285, '2020-11-22 06:19:24', '"grillfuerst.de" <shop@grillfuerst.de>', '"DEBUG"	6324	"2020-11-29 23:02:27.052"	"Error saving the index."
"DEBUG"	8660	"2020-11-29 23:02:28.193"	"Event completed"
"SMTPD"	8660	1304	"2020-11-29 23:02:28.193"	"xx.xxx.xxx.xxx"	"SENT: 235 authenticated."
"SMTPD"	8988	1304	"2020-11-29 23:02:28.209"	"xx.xxx.xxx.xxx"	"RECEIVED: MAIL FROM:<xxx@xxx.de>"
"SMTPD"	8988	1304	"2020-11-29 23:02:28.209"	"xx.xxx.xxx.xxx"	"SENT: 250 OK"
"SMTPD"	4988	1304	"2020-11-29 23:02:28.209"	"xx.xxx.xxx.xxx"	"RECEIVED: RCPT TO:<xxx@xxx.de>"
"SMTPD"	4988	1304	"2020-11-29 23:02:28.209"	"xx.xxx.xxx.xxx"	"SENT: 250 OK"
"SMTPD"	6396	1304	"2020-11-29 23:02:28.224"	"xx.xxx.xxx.xxx"	"SENT: 354 OK, send."
"DEBUG"	6396	"2020-11-29 23:02:28.490"	"Could not retrieve PTR record for IP (false)! xx.xxx.xxx.xxx"
"DEBUG"	6396	"2020-11-29 23:02:28.490"	"Adding task AsynchronousTask to work queue Asynchronous task queue"
"DEBUG"	5444	"2020-11-29 23:02:28.505"	"Executing task AsynchronousTask in work queue Asynchronous task queue"
"SMTPD"	5444	1304	"2020-11-29 23:02:28.521"	"xx.xxx.xxx.xxx"	"SENT: 554 Rejected - Message containing bare LF's."
"DEBUG"	5444	"2020-11-29 23:02:28.521"	"Deleting message file."
"DEBUG"	8660	"2020-11-29 23:02:28.521"	"The read operation failed. Bytes transferred: 0 Remote IP: xx.xxx.xxx.xxx, Session: 1304, Code: 2, Message: End of file"
"DEBUG"	8660	"2020-11-29 23:02:28.521"	"Ending session 1304"```

Any ideas what's going wrong here?
Can not give you the version right now, since I am using "hMailServer-1.0.0-B0-x64" according to file and help dialogue... It might be "hMailServer-5.6.7-B2425" or later.

**Comment by OutbackMatt** [comment]
This should really go to the user forum >> https://www.hmailserver.com/forum/

Version shows in the Admin GUI  >> Status >> server page

**Comment by visualperception** [comment]
Run hmailserver administrator, i.e. run hMailAdmin.exe from wherever you have it located.
then go to following:
settings/protocols/smtp
then select from tabs on right: RFC Compliance.
Then tick: Allow incorrectly formatted line ending.
Then click save and then restart hMailserver and see how you go.
Can't guarantee it will work but it won't do any harm. I have mine ticked since there are so many old mail programs out there that don't comply with current RFC's, And a lot of badly formatted emails from the likes of PHP and Microsoft asp etc etc.

**Comment by FabulousGee** [comment]
> Run hmailserver administrator, i.e. run hMailAdmin.exe from wherever you have it located.
> then go to following:
> settings/protocols/smtp
> then select from tabs on right: RFC Compliance.
> Then tick: Allow incorrectly formatted line ending.
> Then click save and then restart hMailserver and see how you go.
> Can't guarantee it will work but it won't do any harm. I have mine ticked since there are so many old mail programs out there that don't comply with current RFC's, And a lot of badly formatted emails from the likes of PHP and Microsoft asp etc etc.

Thanks for the hint, I know this. But I wanted it to be RFC compliant since we got some rejected mails from other mailservers so we need to keep it clean on our side...



> This should really go to the user forum >> https://www.hmailserver.com/forum/

Is it enough to post a reference? Or is it better to post everything on there again? I want to avoid splitting up the answers nor have them double-posted.

> Version shows in the Admin GUI >> Status >> server page

Nope... Like I said, it is not shown correctly in this version. Don't trust me? Click here... https://abload.de/img/nopevkjlb.png

**Comment by OutbackMatt** [comment]
A link will suffice, but copy and paste text too please.

Where did you get the version that you are running?

**Comment by visualperception** [comment]
> > Run hmailserver administrator, i.e. run hMailAdmin.exe from wherever you have it located.
> > then go to following:
> > settings/protocols/smtp
> > then select from tabs on right: RFC Compliance.
> > Then tick: Allow incorrectly formatted line ending.
> > Then click save and then restart hMailserver and see how you go.
> > Can't guarantee it will work but it won't do any harm. I have mine ticked since there are so many old mail programs out there that don't comply with current RFC's, And a lot of badly formatted emails from the likes of PHP and Microsoft asp etc etc.
> > 
> >Thanks for the hint, I know this. But I wanted it to be RFC compliant since we got some rejected mails from other mailservers so we need to keep it clean on our side...

Well the choice is yours but if you are sending out badly formatted emails and dont tick it you are probably going to lose quite a few incoming emails. You would best off tracing the source (program, script) that is generating these emails and fixing that if you suspect it is one of yours.

**Comment by FabulousGee** [comment]
> A link will suffice, but copy and paste text too please.
> 
> Where did you get the version that you are running?

Will do that later on. The version was out of the forums, I think. Not sure if it was an official release or some modified version (since it was for x64) Sorry for not being more specific, I just don't remember anymore.

> Well the choice is yours but if you are sending out badly formatted emails and dont tick it you are probably going to lose quite a few incoming emails. You would best off tracing the source (program, script) that is generating these emails and fixing that if you suspect it is one of yours.

I am/we are not sending out badly formatted emails right now - since they get blocked by our own hmailserver.
Before I activated the checkbox, they were sent out but got blocked by recipients mailserver. And even if they respond/send badly formatted mails now, they get a rejection note just like we did since the server responds with the error.
So we don't loose anything imho - correct me if I am wrong.
In my opinion it's not a solution to just accept and ignore mistakes - this is more like a workaround.

**Comment by visualperception** [comment]
>the users can not use our webmailer anymore!

There you have it. You are admitting that your own web mailer is not formatting line endings correctly since they only stopped being able send emails when you unticked " Allow incorrectly formatted line endings". I suggest you fix the problem you have with your web mailer and the problem will go away. If you do not fix it you will continue to have problems.

**Comment by FabulousGee** [comment]
I am right here to fix it but now with a workaround, it's not that urgent... They can use Thunderbird or whatever works fine.
It wouldn't be any advantage to have the mails accepted since they get not delivered either way!

Currently we are using RainLoop in the latest version (1.14.0 iirc) but also tried snappymail without success. We had roundcube before but it didn't suit our needs. Do you know good alternatives?

**Comment by visualperception** [comment]
> Currently we are using RainLoop in the latest version (1.14.0 iirc) but also tried snappymail without success. We had roundcube before but it didn't suit our needs. Do you know good alternatives?

What's a good alternative? We don't know your requirements and expectations of a webmailer. You need to think about whether you need a mobile app for your users however many they are. And also whether you need groupware or something simplistic. 
There is also the the security level and spec you want for it etc etc etc and  this is not the place to discuss it.

**Comment by the-djmaze** [comment]
hMail server decodes the quoted-printable text/plain and then runs the check.
=0A is not CRLF, it is LF.
=0D=0A is CRLF
hMail should not run the check after decoding because it can be on purpose.

**Comment by the-djmaze** [comment]
Although our webmail app is correct, i made changes to prevent the hMailserver bug.
You should fix your code to prevent valid mime messages to be blocked.

**Comment by visualperception** [comment]
> hMail server decodes the quoted-printable text/plain and then runs the check.
> =0A is not CRLF, it is LF.
> =0D=0A is CRLF
> hMail should not run the check after decoding because it can be on purpose.

Then you should start a new issue explaining exactly what you think the problem is now that you have narrowed it down and someone who knows github and C/C++ may have a look at it. Unfortunately that's not me. @RvdHout is your best bet.

**Comment by the-djmaze** [comment]
New issue not really needed because @FabulousGee does explain is very well.

Problem should be here:
https://github.com/hmailserver/hmailserver/blob/master/hmailserver/source/Server/SMTP/SMTPConnection.cpp#L1279

Looking at it, it does have bugs in reading pBuffer
What if ReadChunk contains 'data\r' and the next chunk '\nmore data'.
And why skip the first bytes in pBuffer chunk?

Outside the while loop there should be a char to hold previous match.

**Comment by visualperception** [comment]
> New issue not really needed because @FabulousGee does explain is very well.
> 
Fair enough. Lets hope somone picks it up. Often when some light is shed on real problem towards end of lengthy issue it doesn't get picked up by those who know how to fix it.

**Comment by FabulousGee** [comment]
Thanks @the-djmaze for the fast workaround - works like a charm!

For the "main fix" on hmailserver-side I re-ran the test with the exact same email. This is what I logged now:
```
TAG1 CAPABILITY
TAG2 LOGIN "xxx.xxx" "xxx"
EHLO mail.xxx.de
AUTH LOGIN
xxx=
xxx=
MAIL FROM:<xxx@xxx.de>
RCPT TO:<xxx@xxx.de>
DATA
MIME-Version: 1.0
Date: Mon, 30 Nov 2020 14:56:01 +0000
Content-Type: multipart/alternative;
 boundary="--=_RainLoop_453_615729926.1606748161"
From: "xxx - xxx"
 <xxx@xxx.de>
Message-ID: <xxx@xxx.de>
Subject: TESTemail
To: xxx@xxx.de
X-Originating-IP: xxx.xx.xx.xxx


----=_RainLoop_453_615729926.1606748161
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: quoted-printable

Das ist nur ein Test

ein Test
mit verschiedenen Umbr=C3=BCchen

	Testtext 12345
123421345
12345
13245
32145
asd
g
asdfg
sdffg

----=_RainLoop_453_615729926.1606748161
Content-Type: text/html; charset="utf-8"
Content-Transfer-Encoding: quoted-printable

<!DOCTYPE html><html><head><meta http-equiv=3D"Content-Type" content=3D"t=
ext/html; charset=3Dutf-8" /></head><body><div data-html-editor-font-wrap=
per=3D"true" style=3D"font-family: arial, sans-serif; font-size: 13px;"><=
div><div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><d=
iv><div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><di=
v><div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><div=
><div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><div>=
<div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><div><=
div><div style=3D"font-family: arial, sans-serif;font-size: 13px"><p>Das =
ist nur ein Test<br><br>ein Test<br>mit verschiedenen Umbr=C3=BCchen</p><=
p>Testtext 12345<br>123421345<br>12345<br>13245<br>32145<br>asd<br>g<br>a=
sdfg<br>sdffg</p><signature></signature></div></div></div></div></div></d=
iv></div></div></div></div></div></div></div></div></div></div></div></di=
v></div></body></html>

----=_RainLoop_453_615729926.1606748161--
.
QUIT
TAG3 APPEND "Sent" (\Seen) {1857}

TAG4 LOGOUT
```

So it really seems to be a false assumption when checking for LF/CRLF with the encoding/decoding

**Comment by SorenRR** [comment]
I'd like to thank @FabulousGee for pointing out a possible BUG in our current **_ALPHA_** release of hMailServer 5.7 x64.

**_ALPHA_** as in "yes there may be tons of bugs but we do a limited release anyway" version.

To @the-djmaze ... Don't get all worked up about this. If you want to certify RainLoop with hMailServer then please use our official release (5.6.7 B2425) or ask RvdH to provide you with the latest "community" edition of 5.6.8 B2505.27 .
5.6.8 B2505.27 is more a "community" release - and a pre-study to 5.7 thus all the new features and script triggers in 5.6.8 B2505.27 is ported to 5.7 x64.

I just compared the "SMTPConnections.cpp" linked to by @the-djmaze with the official release version of said file and they do NOT compare.

**Comment by the-djmaze** [comment]
@SorenRR i'm not from RainLoop.
I just develop a secure fork, named SnappyMail, for my servers.

Great you've compared the files.
But which hMailserver files should i check and submit a patch?
I'm just a multi(programming)lingual using Linux.

**Comment by SorenRR** [comment]
Good question @the-djmaze, my "production" server is still 32-bit so I'm stuck with 5.6.8. Actually It's a custom version of 5.6.8 as I have added/changed a few bits and pieces for my own use ;-)

I believe "The Man" for 5.7 codechanges is @RvdHout as @martinknafve is not here that much.

**Comment by RvdHout** [comment]
@the-djmaze 
It would be helpful if you could confirm it is only a issue in the ALPHA (x64) release and not a issue in the stable 5.6.8-B* x86 release

@SorenRR 
True they don't compare....but what did you expect? There is done variable conversion for x64

What is the function you guys think has the error in? CheckLineEndings_()
https://github.com/hmailserver/hmailserver/blob/master/hmailserver/source/Server/SMTP/SMTPConnection.cpp#L1289

**Comment by the-djmaze** [comment]
@RvdHout yes, but there seems to be 2 bugs.
1. CheckLineEndings_() is executed AFTER the mail is decoded. So it checks if if the ".txt" uses CRLF instead of the mime encoded version.
2. CheckLineEndings_() has a small bug where it only looks for CRLF in the current buffer but fails when the CR is the last in the buffer and LF is the first character in the next buffer, and `for (size_t i = 3` skips the first 3 bytes in buffer.

**Comment by RvdHout** [comment]
@the-djmaze 
I doubt hmailserver does any decoding, it seems to read file from disk inside CheckLineEndings_(), i don't see anywhere it tries to decode anything

I noticed the skipping of the first 3 bytes yes, weird...

`         if (iBufferSize >= 3)
         {
            for (size_t i = 3; i < iBufferSize - 3; i++)
            {
}}`

@martinknafve, can you give us some insight?

**Comment by the-djmaze** [comment]
@RvdHout
If hmailserver doesn't decode, then why does CheckLineEndings_() say it is a faulty e-mail when using:

`----=_RainLoop_472_915006015.1606687346
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: quoted-printable

Das ist nur ein Test=0A=0Aein Test=0Amit verschiedenen Umbr=C3=BCchen=0A=
=0A	Testtext 12345=0A123421345=0A12345=0A13245=0A32145=0Aasd=0Ag=0Aasdfg=
=0Asdffg`

So it seems hmailserver does decode the =0A and then CheckLineEndings_().

**Comment by FabulousGee** [comment]
Any news on this?

We are experiencing the same behaviour on our ticketing system (osTicket)

**Comment by FabulousGee** [comment]
Still an issue - no one going for it?
